packages/devtools_app/test/support/mocks.dart

```diff
@@ -15,8 +19,269 @@ class MockServiceManager extends Mock implements ServiceConnectionManager {
+  final Set<String> _serviceExtensions = Set<String>();
+
+  /// All service extensions that are currently enabled.
+  final Map<String, ServiceExtensionState> _enabledServiceExtensions =
+      <String, ServiceExtensionState>{};
+
+  /// Temporarily stores service extensions that we need to add. We should not
+  /// add extensions until the first frame event has been received
+  /// [_firstFrameEventReceived].
+  // ignore: prefer_collection_literals
+  final Set<String> _pendingServiceExtensions = Set<String>();
+
+  @override
+  var extensionStatesUpdated = Completer<void>();
+
+  /// Hook to simulate receiving the first frame event.
+  ///
+  /// Service extensions are only reported once a frame has been received.
+  void fakeFrame() async {
+    await _onFrameEventReceived();
+  }
+
+  Map<String, dynamic> extensionValueOnDevice = {};
+
+  /// Hook for tests to call to simulate adding a service extension.
+  Future<void> fakeAddServiceExtension(String name) async {
+    if (_firstFrameEventReceived) {
+      assert(_pendingServiceExtensions.isEmpty);
+      await _addServiceExtension(name);
+    } else {
+      _pendingServiceExtensions.add(name);
+    }
+  }
+
+  /// Hook for tests to call to fake changing the state of a service
+  /// extension.
+  void fakeServiceExtensionStateChanged(
+      final String name, String valueFromJson) async {
+    final extension = extensions.serviceExtensionsWhitelist[name];
+    if (extension != null) {
+      final dynamic value = _getExtensionValueFromJson(name, valueFromJson);
+
+      final enabled =
+          extension is extensions.ToggleableServiceExtensionDescription
+              ? value == extension.enabledValue
+              // For extensions that have more than two states
+              // (enabled / disabled), we will always consider them to be
+              // enabled with the current value.
+              : true;
+
+      await setServiceExtensionState(
+        name,
+        enabled,
+        value,
+        callExtension: false,
+      );
+    }
+  }
+
+  dynamic _getExtensionValueFromJson(String name, String valueFromJson) {
+    final expectedValueType =
+        extensions.serviceExtensionsWhitelist[name].values.first.runtimeType;
+    switch (expectedValueType) {
+      case bool:
+        return valueFromJson == 'true' ? true : false;
+      case int:
+      case double:
+        return num.parse(valueFromJson);
+      default:
+        return valueFromJson;
+    }
+  }
+
+  Future<void> _onFrameEventReceived() async {
+    if (_firstFrameEventReceived) {
+      // The first frame event was already received.
+      return;
+    }
+    _firstFrameEventReceived = true;
+
+    for (String extension in _pendingServiceExtensions) {
+      await _addServiceExtension(extension);
+    }
+    extensionStatesUpdated.complete();
+    _pendingServiceExtensions.clear();
+  }
+
+  Future<void> _addServiceExtension(String name) async {
+    final StreamController<bool> streamController =
```

Comment:
remove type? Or is this preferred in flutter style?

Reply:
Inside of a method body, I agree.  The left-hand type isn't giving much useful information here.

Reply:
removed. this is just copy and paste code of the fraction of ServiceExtensionManager required to create a plausible fake version.